csharp/src/ProtocolBuffers/IMessage.cs

```diff
@@ -93,4 +93,22 @@ public interface IMessage<T> : IMessage where T : IMessage<T>
         /// <param name="message">The message to merge with this one. Must not be null.</param>
         void MergeFrom(T message);
     }
+
+    /// <summary>
+    /// Generic interface for a deeply cloneable type.
+    /// <summary>
+    /// <remarks>
+    /// In practice, all generated messages implement this interface.
+    /// However, due to the type constraint on <c>T</c> in <see cref="IMessage{T}"/>,
+    /// it is simpler to keep this as a separate interface.
+    /// </remarks>
+    /// <typeparam name="T">The type itself, returned by the <see cref="Clone"/> method.</typeparam>
+    public interface IDeepCloneable<T>

```

Comment:
I had some doubts about the name IDeepCloneable, but when rethinking it, it's probably fine to be more descriptive.